pkg/cloudprovider/providers/azure/azure_wrap.go

```diff
@@ -173,3 +139,128 @@ func (az *Cloud) getSubnet(virtualNetworkName string, subnetName string) (subnet
+	if err != nil {
+		return nsg, err
+	}
+
+	if securityGroup == nil {
+		return nsg, fmt.Errorf("nsg %q not found", az.SecurityGroupName)
+	}
+
+	return *(securityGroup.(*network.SecurityGroup)), nil
+}
+
+func (az *Cloud) newVMCache() *timedCache {
+	getter := func(key string) (interface{}, error) {
+		vm, err := az.VirtualMachinesClient.Get(az.ResourceGroup, key, compute.InstanceView)
+		exists, realErr := checkResourceExistsFromError(err)
+		if realErr != nil {
+			return nil, realErr
+		}
+
+		if !exists {
+			return nil, nil
+		}
+
+		return &vm, nil
+	}
+
+	lister := func() (map[string]interface{}, error) {
+		allNodes := map[string]interface{}{}
+
+		result, err := az.VirtualMachinesClient.List(az.ResourceGroup)
+		if err != nil {
+			return nil, err
+		}
+		moreResults := (result.Value != nil && len(*result.Value) > 0)
+		if moreResults {
+			for idx := range *result.Value {
+				vm := (*result.Value)[idx]
+				allNodes[*vm.Name] = &vm
+			}
+			moreResults = false
+			result, err = az.VirtualMachinesClient.ListNextResults(az.ResourceGroup, result)
+			if err != nil {
+				return nil, err
+			}
+			moreResults = (result.Value != nil && len(*result.Value) > 0)
+		}
+
+		return allNodes, nil
+	}
+
+	return newTimedcache(vmCacheTTL, getter, lister)
+}
+
+func (az *Cloud) newLBCache() *timedCache {
+	getter := func(key string) (interface{}, error) {
+		lb, err := az.LoadBalancerClient.Get(az.ResourceGroup, key, "")
+		exists, realErr := checkResourceExistsFromError(err)
+		if realErr != nil {
+			return nil, realErr
+		}
+
+		if !exists {
+			return nil, nil
+		}
+
+		return &lb, nil
+	}
+
+	lister := func() (map[string]interface{}, error) {
+		allLBs := map[string]interface{}{}
+
+		result, err := az.LoadBalancerClient.List(az.ResourceGroup)
+		if err != nil {
+			return nil, err
+		}
+		moreResults := (result.Value != nil && len(*result.Value) > 0)
+		if moreResults {
+			for idx := range *result.Value {
+				lb := (*result.Value)[idx]
+				allLBs[*lb.Name] = &lb
+			}
+			moreResults = false
+			result, err = az.LoadBalancerClient.ListNextResults(az.ResourceGroup, result)
+			if err != nil {
+				return nil, err
+			}
+			moreResults = (result.Value != nil && len(*result.Value) > 0)
+		}
+
+		return allLBs, nil
+	}
+
+	return newTimedcache(vmCacheTTL, getter, lister)
+}
+
+func (az *Cloud) newNSGCache() *timedCache {
+	getter := func(key string) (interface{}, error) {
+		lb, err := az.SecurityGroupsClient.Get(az.ResourceGroup, key, "")
+		exists, realErr := checkResourceExistsFromError(err)
+		if realErr != nil {
+			return nil, realErr
+		}
+
+		if !exists {
+			return nil, nil
+		}
+
+		return &lb, nil
+	}
+
+	lister := func() (map[string]interface{}, error) {
+		// Only one security group is used.
+		lb, err := getter(az.SecurityGroupName)
+		if err != nil {
+			return nil, err
+		}
+
+		return map[string]interface{}{az.SecurityGroupName: lb}, nil
+	}
+
+	return newTimedcache(lbCacheTTL, getter, lister)
```

Comment:
NSG Shares lbCacheTTL? Should better be explicitly called out in const definition.